Fails to build against VTK v9.4 (released on 2024-11-22 + has some breaking changes)

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Building 1.0.0 with vtk-9.4 fails, with vtk-9.3 it works, looks like [9.4 has some breaking changes](https://docs.vtk.org/en/latest/release_details/9.4.html).

````
 39%] Building CXX object src/3rdParty/salomesmesh/CMakeFiles/SMDS.dir/src/SMDS/SMDS_VolumeOfNodes.cpp.o
/usr/src/t2-src/src.freecad.241210.214008.18573/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp: In member function 'void SMDS_UnstructuredGrid::compactGrid(std::vector<int, std::allocator<int> >&, int, std::vector<int, std::allocator<int> >&, int)':
/usr/src/t2-src/src.freecad.241210.214008.18573/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp:268:54: error: 'class vtkCellArray' has no member named 'GetValue'
  268 |                int oldFaceLoc = this->FaceLocations->GetValue(i);
      |                                                      ^~~~~~~~
/usr/src/t2-src/src.freecad.241210.214008.18573/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp:269:46: error: 'class vtkCellArray' has no member named 'GetValue'
  269 |                int nCellFaces = this->Faces->GetValue(oldFaceLoc++);
      |                                              ^~~~~~~~
/usr/src/t2-src/src.freecad.241210.214008.18573/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp:273:50: error: 'class vtkCellArray' has no member named 'GetValue'
  273 |                    int nptsInFace = this->Faces->GetValue(oldFaceLoc++);
      |                                                  ^~~~~~~~
/usr/src/t2-src/src.freecad.241210.214008.18573/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp:277:49: error: 'class vtkCellArray' has no member named 'GetValue'
  277 |                        int oldpt = this->Faces->GetValue(oldFaceLoc++);
      |                                                 ^~~~~~~~
/usr/src/t2-src/src.freecad.241210.214008.18573/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp:295:19: error: no matching function for call to 'SMDS_UnstructuredGrid::SetCells(vtkUnsignedCharArray*&, vtkIdTypeArray*&, vtkCellArray*&, vtkSmartPointer<vtkCellArray>&, vtkSmartPointer<vtkCellArray>&)'
  295 |     this->SetCells(newTypes, newLocations, newConnectivity, FaceLocations, Faces);
````

### Full version info

```shell
OS: Linux 6.12.1-t2 (xcb)
Architecture: x86_64
Version: 1.0.0.39109 (Git)
Build type: Release
Branch: (HEAD detached at 1.0.0)
Hash: 2fcc5317fe3aee96ca73475986a577719fc78e20
Python 3.12.3, Qt 6.8.1, Coin 4.0.3, Vtk 9.3.1, OCC 7.8.1
Locale: German/Germany (de_DE)
Stylesheet/Theme/QtStyle: FreeCAD Light.qss/FreeCAD Light/
```


### Subproject(s) affected?

None

### Anything else?

Build with vtk-9.3 works fine (shown in about info), with 9.4 not.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

I cherry-picked the commit ontop of [main](https://github.com/anrub/FreeCAD) and built it against VTK 9.4.0, and get the error, but unrelated to the change, it also occurs on main with VTK-9.4.0
` SMDS_UnstructuredGrid.cpp:1032:25: error: no matching function for call to 'vtkCellLinks::BuildLinks(SMDS_UnstructuredGrid*)''` 

Line 3797 in the full build log [freecad.log.txt](https://github.com/user-attachments/files/18210411/freecad.log.txt)

```
/usr/src/t2-src/src.freecad.241220.123528.21095/FreeCAD.git/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp: In member function 'void SMDS_UnstructuredGrid::BuildLinks()':
/usr/src/t2-src/src.freecad.241220.123528.21095/FreeCAD.git/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp:1032:25: error: no matching function for call to 'vtkCellLinks::BuildLinks(SMDS_UnstructuredGrid*)'
 1032 |   GetLinks()->BuildLinks(this);
      |   ~~~~~~~~~~~~~~~~~~~~~~^~~~~~
In file included from /usr/src/t2-src/src.freecad.241220.123528.21095/FreeCAD.git/src/3rdParty/salomesmesh/inc/SMDS_UnstructuredGrid.hxx:30:
/usr/include/vtk-9.4/vtkCellLinks.h:64:8: note: candidate: 'virtual void vtkCellLinks::BuildLinks()'
   64 |   void BuildLinks() override;
      |        ^~~~~~~~~~
/usr/include/vtk-9.4/vtkCellLinks.h:64:8: note:   candidate expects 0 arguments, 1 provided
[ 40%] Building CXX object src/3rdParty/salomesmesh/CMakeFiles/SMDS.dir/src/SMDS/SMDS_VertexPosition.cpp.o
make[2]: *** [src/3rdParty/salomesmesh/CMakeFiles/SMDS.dir/build.make:474: src/3rdParty/salomesmesh/CMakeFiles/SMDS.dir/src/SMDS/SMDS_UnstructuredGrid.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

--- a/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp
+++ b/src/3rdParty/salomesmesh/src/SMDS/SMDS_UnstructuredGrid.cpp
@@ -32,6 +32,7 @@
 #include <vtkDoubleArray.h>
 #include <vtkIdTypeArray.h>
 #include <vtkUnsignedCharArray.h>
+#include <vtkVersionMacros.h>
 
 #include <list>
 #include <climits>
@@ -249,14 +250,16 @@ void SMDS_UnstructuredGrid::compactGrid(std::vector<int>& idNodesOldToNew, int n
     }
   }
 
-  if (this->FaceLocations)
+  vtkIdTypeArray* thisFaceLocations = GetFaceLocations();
+  vtkIdTypeArray* thisFaces = GetFaces();
+  if (thisFaceLocations)
     {
       vtkIdTypeArray *newFaceLocations = vtkIdTypeArray::New();
       newFaceLocations->Initialize();
       newFaceLocations->Allocate(newTypes->GetSize());
       vtkIdTypeArray *newFaces = vtkIdTypeArray::New();
       newFaces->Initialize();
-      newFaces->Allocate(this->Faces->GetSize());
+      newFaces->Allocate(thisFaces->GetSize());
       for (int i = 0; i < oldCellSize; i++)
         {
           if (this->Types->GetValue(i) == VTK_EMPTY_CELL)
@@ -265,16 +268,16 @@ void SMDS_UnstructuredGrid::compactGrid(std::vector<int>& idNodesOldToNew, int n
           if (newTypes->GetValue(newCellId) == VTK_POLYHEDRON)
             {
                newFaceLocations->InsertNextValue(newFaces->GetMaxId()+1);
-               int oldFaceLoc = this->FaceLocations->GetValue(i);
-               int nCellFaces = this->Faces->GetValue(oldFaceLoc++);
+               int oldFaceLoc = thisFaceLocations->GetValue(i);
+               int nCellFaces = thisFaces->GetValue(oldFaceLoc++);
                newFaces->InsertNextValue(nCellFaces);
                for (int n=0; n<nCellFaces; n++)
                  {
-                   int nptsInFace = this->Faces->GetValue(oldFaceLoc++);
+                   int nptsInFace = thisFaces->GetValue(oldFaceLoc++);
                    newFaces->InsertNextValue(nptsInFace);
                    for (int k=0; k<nptsInFace; k++)
                      {
-                       int oldpt = this->Faces->GetValue(oldFaceLoc++);
+                       int oldpt = thisFaces->GetValue(oldFaceLoc++);
                        newFaces->InsertNextValue(idNodesOldToNew[oldpt]);
                      }
                  }
@@ -292,7 +295,7 @@ void SMDS_UnstructuredGrid::compactGrid(std::vector<int>& idNodesOldToNew, int n
     }
   else
   {
-    this->SetCells(newTypes, newLocations, newConnectivity, FaceLocations, Faces);
+    this->SetCells(newTypes, newLocations, newConnectivity, thisFaceLocations, thisFaces);
   }
 
   newPoints->Delete();

